### DISCUSSION TITLE
싱글톤 선언 또는 static 사용방법

### DISCUSSION THREAD

**Discussion by JeongDongBaek** [naverz/zepeto-studio-kor#215] [5. Q&A]
올바른 방법이 있을까요?

해당 클래스에 static을 붙일 수 없는 구조로 보입니다.

변수나 함수에 static으로 감싸고 다른 클래스에서 호출하려고 하면 property null 에러가 나옵니다.

기존의 public static instance를 선언하고 static빼고 해보았지만 똑같습니다.

SendMessage는 쓰고싶지않습니다.

인스턴스화한다는 가정으로 기존 c#처럼 싱글톤하려면 어떻게 해야할까요?

**Comment by moondory77-naverz** [comment]
안녕하세요. 
제페토 스크립트는 기본적으로 TypeScript 방식의 개발루틴을 가지고 있습니다.

Class 자체에 static 키워드를 붙일 수 없음은 C#과 TypeScript의 차이에서 기인합니다.

다만, TypeScript는 프로퍼티와 메소드에 static 키워드를 사용할 수 있으며
매우 다양한 방법을 통해, 기존 유니티 C# 스크립트와 동일한 로직을 구현할 수 있습니다.

말씀주신, 싱글톤 패턴 또한 충분히 사용가능합니다.

예를 들어 MonoBehaviour에 상응하는 ZepetoScriptBehaviour를 싱글톤 패턴으로 사용한다고 했을 때,
다음과 같은 방식으로 사용할 수 있습니다.

```
export default class CustomSingetonBehaviour extends ZepetoScriptBehaviour  {

    private static Instance: CustomSingetonBehaviour;
    public static GetInstance(): CustomSingetonBehaviour {

        if (!CustomSingetonBehaviour.Instance) {

            var _obj = new GameObject("CustomSingleton");
            GameObject.DontDestroyOnLoad(_obj);
            CustomSingetonBehaviour.Instance = _obj.AddComponent<CustomSingetonBehaviour>();
        }
        return CustomSingetonBehaviour.Instance;
    }

    public myProp1: number;
   
    public myProp2: string;
    
    ...
}
```

도움이 될 수 있다면 좋겠습니다.
감사합니다.
